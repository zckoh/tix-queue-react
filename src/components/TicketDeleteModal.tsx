import moment from "moment";
import { useState } from "react";
import { Button, Modal } from "react-bootstrap";
import { v4 as uuidv4 } from "uuid";

import { deleteTicketById } from "../services";
import { Notification } from "../types";

interface TicketDeleteModalPropTypes {
  show: Boolean;
  ticketTitle: String;
  ticketId: Number;
  handleClose: () => void;
  handleNotification: (notification: Notification) => void;
  handleSuccessDelete: () => void;
}

const TicketDeleteModal = (props: TicketDeleteModalPropTypes) => {
  const { show, ticketTitle, ticketId, handleClose, handleNotification, handleSuccessDelete } = props;

  const [isLoading, setLoading] = useState(false);
  const [isDeleted, setDeleted] = useState(false);
  const [isError, setError] = useState(false);

  const deleteTicket = () => {
    setLoading(true);

    // Post new ticket data
    deleteTicketById(ticketId)
      .then(() => {
        setLoading(false);
        setDeleted(true);
        handleNotification({
          notificationId: uuidv4(),
          message: `✔️ - Ticket "${ticketTitle}" deleted successfully.`,
          status: "success",
          show: true,
          createdAt: moment(),
        });
        closeAllModal();
      })
      .catch(() => {
        setLoading(false);
        setError(true);
        handleNotification({
          notificationId: uuidv4(),
          message: "❌ - Failed to delete ticket.",
          status: "failed",
          show: true,
          createdAt: moment(),
        });
      });
  };

  const closeAllModal = () => {
    handleClose();
    handleSuccessDelete();
  };

  return (
    <Modal
      show={show}
      onHide={() => handleClose()}
      className="modal-dark"
      animation={true}
      backdropClassName="delete-modal-backdrop"
    >
      <Modal.Header closeButton>
        <Modal.Title>Delete Ticket</Modal.Title>
      </Modal.Header>
      <Modal.Body>
        You are about to delete the following ticket:
        <br />- {ticketTitle}
      </Modal.Body>
      <Modal.Footer>
        <Button variant="secondary" onClick={handleClose}>
          {isDeleted ? `Close` : `Cancel`}
        </Button>

        <Button onClick={deleteTicket} active={isLoading} disabled={isDeleted} variant="danger">
          {isDeleted
            ? `Ticket Deleted`
            : isLoading
            ? `Deleting...`
            : isError
            ? `Failed to delete ticket`
            : // Default
              `Delete Ticket`}
        </Button>
      </Modal.Footer>
    </Modal>
  );
};

export default TicketDeleteModal;
