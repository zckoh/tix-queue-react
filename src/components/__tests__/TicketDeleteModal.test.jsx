import { mount } from "enzyme";
import { Button, Modal } from "react-bootstrap";
import { waitFor } from "@testing-library/react";
import TicketDeleteModal from "../TicketDeleteModal";
import { enableFetchMocks } from "jest-fetch-mock";
enableFetchMocks();

beforeEach(() => {
  fetch.resetMocks();
});

describe("<TicketDeleteModal />", () => {
  const mockHandleClose = jest.fn();
  const mockHandleNotification = jest.fn();
  const mockHandleSuccessDelete = jest.fn();

  const inputTicketTitle = "This is a ticket title";
  const inputTicketId = 10;

  const wrapper = mount(
    <TicketDeleteModal
      show={false}
      handleClose={mockHandleClose}
      handleSuccessDelete={mockHandleSuccessDelete}
      ticketTitle={inputTicketTitle}
      ticketId={inputTicketId}
      handleNotification={mockHandleNotification}
    />,
  );

  it(`should correctly renders component`, () => {
    expect(wrapper).toMatchSnapshot();
  });

  it(`should display ticket information when modal is shown`, () => {
    // Show Modal
    wrapper.setProps({ show: true });
    wrapper.update();
    const ticketDeleteModal = wrapper.find(Modal).first();
    expect(ticketDeleteModal.props().show).toBeTruthy();

    const modalBody = ticketDeleteModal.find(Modal.Body);
    expect(modalBody.props().children).toContain(inputTicketTitle);
  });

  it(`should call all handle props function when click delete button`, async () => {
    fetch.mockResponseOnce(`Ticket with id - ${inputTicketId} is deleted`);

    const deleteButton = wrapper.find(Button).findWhere((n) => n.props().variant === "danger");
    deleteButton.simulate("click");

    await waitFor(() => {
      expect(mockHandleClose).toHaveBeenCalledTimes(1);
      expect(mockHandleNotification).toHaveBeenCalledTimes(1);
      expect(mockHandleSuccessDelete).toHaveBeenCalledTimes(1);
    });
  });
});
