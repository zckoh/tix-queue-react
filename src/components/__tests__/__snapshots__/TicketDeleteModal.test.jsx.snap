// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TicketDeleteModal /> should correctly renders component 1`] = `ReactWrapper {}`;
